src/matcher/matcher.go

```diff
@@ -0,0 +1,104 @@
+package matcher
+
+import (
+	"bufio"
+	"regexp"
+	"strings"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+)
+
+type matchResult int
+
+const (
+	noMatch matchResult = iota
+	exclude
+	include
+)
+
+//Matcher groups the rules to validate the service name
+type Matcher struct {
+	patterns []pattern
+}
+type pattern struct {
+	exclude bool
+	regex   string
+}
+
+// Match returns true if the string matches one of the patterns
+// within one level of precedence, the last matching pattern decides the outcome
+func (m *Matcher) Match(s string) bool {
+	n := len(m.patterns)
+	for i := n - 1; i >= 0; i-- {
+		if match := m.patterns[i].match(s); match > noMatch {
+			return match == include
+		}
+	}
+	return false
+}
+func (p pattern) match(s string) matchResult {
+	match, err := regexp.MatchString(p.regex, s)
+	if err != nil {
+		log.Warn(err.Error())
```

Comment:
Make sense to add more detail on the log? such as the service is trying to match and/or the regex

Reply:
yes and actually i just realize that the regex compilation should be done once inside the New func, I will udpate that and add the proper log there , thanks!